README.md
# ComicBook

漫画爬虫、漫画下载工具，支持腾讯漫画、哔哩哔哩漫画、有妖气漫画、快看漫画、漫画柜等站点

[漫画源收录情况](https://github.com/lossme/ComicBook/projects/1)

尊重版权，请支持正版，通过本工具下载或生成的资源**禁止传播分享！禁止利用本项目进行商业活动！**

## 本项目特点

- [x] 漫画批量下载
- [x] 分目录按章节保存
- [x] 支持多个漫画源，易于扩展
- [x] 支持登录
- [x] 支持生成pdf
- [x] 支持发送到邮箱
- [x] 支持设置代理
- [x] 支持API调用 [API-README](API-README.md)

## 安装/升级步骤

### 方式一（推荐）

```sh
# 在线安装/升级
# 由于网络环境不好导致安装失败 可以搜索关键词尝试解决: github host 修改
python3 -m pip install -U git+https://github.com/lossme/ComicBook

# 查看帮助
python3 -m onepiece --help
```

### 方式二（源码安装）

```sh
# clone项目 或从这里下载最新的代码并解压 https://github.com/lossme/ComicBook/releases
git clone git@github.com:lossme/ComicBook.git
# 切换工作目录
cd ComicBook
# 安装
python3 setup.py clean --all install
# 查看帮助
python3 -m onepiece --help
```

如果在使用过程中，发现问题可以先更新代码再试下，说不定已经修复了。

Star防止走丢，欢迎大家提建议和issue，本项目持续更新。

## 常规使用

从腾讯漫画下载：

- 从URL下载漫画的最新一集: `python3 -m onepiece --url "http://ac.qq.com/Comic/ComicInfo/id/505430"`
- 下载漫画 id=505430 最新一集: `python3 -m onepiece --site=qq --comicid=505430`
- 下载所有章节: `python3 -m onepiece --site=qq --comicid=505430 --all`
- 下载第800集: `python3 -m onepiece --site=qq --comicid=505430 --chapter=800`
- 下载倒数第二集: `python3 -m onepiece --site=qq --comicid=505430 --chapter=-2`
- 下载1到5集,7集，9到10集: `python3 -m onepiece --site=qq --comicid=505430 --chapter=1-5,7,9-10`
- 拼接成长图: `python3 -m onepiece --site=qq --comicid=505430 --single-image --quality 95 --max-height 20000`
- 设置代理: `python3 -m onepiece --site=qq --comicid=505430 --proxy "socks5://127.0.0.1:1080"`
- 自定义保存目录: `python3 -m onepiece --site=qq --comicid=505430 --output MyComicBook`
- 下载单行本: `python3 -m onepiece --site=manhuagui --comicid=1128 --ext-name=单行本 --chapter=-1`
- 通过指定的URL文件列表批量下载[【文件示例】](https://github.com/lossme/ComicBook/blob/master/test/test-url-file.txt)：`python3 -m onepiece --url-file test/test-url-file.txt`
- 若不清楚或不记得comicid，可以使用名字来搜索：`python3 -m onepiece --site=qq --name=海贼`
- 推送到邮箱:
```sh
# 注意: 发送到邮箱需预先配置好信息
# 配样例参照 https://github.com/lossme/ComicBook/blob/master/config.ini.example
# 并根据实际情况调整，将配置文件保存为 config.ini
python3 -m onepiece --site=qq --comicid=505430 --pdf --mail --config config.ini
```
- 生成pdf文件:
```sh
# 生成pdf文件需要额外安装依赖，需要先执行 python3 -m pip install img2pdf 或 python3 -m pip install reportlab
python3 -m onepiece --site=qq --comicid=505430 --pdf
```
- 从其它站点下载，注意不同站点的comicid区别
```sh
# 从哔哩哔哩漫画下载:
python3 -m onepiece --site=bilibili --comicid=mc24742 --chapter=1
# 从有妖气漫画下载:
python3 -m onepiece --site=u17 --comicid=195 --chapter=1
```

### 关于登录

1. 安装selenium: `python3 -m pip install selenium`
2. 安装chrome浏览器，或其它浏览器
3. [下载chromedriver](https://chromedriver.chromium.org/downloads)，或其它浏览器的driver
4. 登录，并将cookies保存在本地（保存登录状态）
```sh
# 在弹出的浏览器上完成登录。若登录完浏览器没自动关闭，可以手动把浏览器关了
python3 -m onepiece --site=qq --comicid=505430 --chapter=-1 \
  --login \
  --driver-path="driver路径" \
  --driver-type="Chrome" \
  --cookies-path="data/cookies/qq.json"
```

### 高级批量下载

```sh
# 有些站点不一定支持，其它的通用参数也适用，可自行组合
# 下载最近更新页面的1到10页 所有漫画的最新一集
python3 -m onepiece --site=nvshens --latest-all --latest-page=1-10

# 展示支持的标签
python3 -m onepiece --site=nvshens --show-tags

# 下载标签搜索结果页面的1到10页 所有漫画的全集
python3 -m onepiece --site=nvshens --tag-all --tag=女神 --tag-page=1-10 --all
```

### 高级配置

#### 通过环境变量配置默认参数
```sh
# 可以将以下命令添加到 ~/.bashrc 或 ~/.zshrc 文件末尾
# 配置默认下载目录
export ONEPIECE_DOWNLOAD_DIR="/home/key/Downloads/MyComicBook"

# 配置默认的邮件配置文件路径
export ONEPIECE_MAIL_CONFIG_FILE="/home/key/MyConfig/config.ini"

# 配置默认的 webdriver
export ONEPIECE_DRIVER_TYPE="Chrome"
export ONEPIECE_DRIVER_PATH="/home/key/data/chromedriver_win32/chromedriver.exe"

# 配置默认的cookies存放目录
export ONEPIECE_COOKIES_DIR="/home/key/data/cookies"

# 配置默认的全局代理地址
export ONEPIECE_PROXY="socks5://127.0.0.1:1082"
# 配置腾讯漫画下载代理地址
export ONEPIECE_PROXY_QQ="socks5://127.0.0.1:1082"
# 配置其它站点下载代理地址
export ONEPIECE_PROXY_XXXX="socks5://127.0.0.1:1082"

# 配置默认的长图质量
export DEFAULT_QUALITY=95
# 配置默认的长图最大高度
export DEFAULT_MAX_HEIGHT=20000
```


CHANGELOG.md
**unreleased**
**v0.3.13**
- 添加漫画源 177漫画、18H漫画区
- 添加图集：Nsfwpicx、宅男女神、秀人网
- 根据最近更新页、标签页批量下载
- 生成长图优化等

**v0.3.12**
- 添加漫画源：土豪漫画网、漫画160、古风漫画网、36漫画网
- 支持给每个站点设置单独的默认代理
- 一些问题修复

**v0.3.11**
- 添加漫画源：COCO漫画、DM5、漫画DB
- 支持通过环境变量设置默认配置
- 支持当单行本、番外篇等扩展漫画下载

**v0.3.10**
- 优化安装方式，提供`setup.py`安装脚本
- webp图片校验优化
- 自动登录逻辑优化，降低各站点登录适配门槛
- api接口支持登录校验

**v0.3.9**
- 添加漫画源：acg456、mh1234、77mh、dmzj
- 支持对18comic的图片重组
- 支持生成zip文件
- 兼容不同版本的pdf生成工具

**v0.3.8**
- 支持通过API管理cookies
- 支持通过API配置代理配置
- fix 腾讯漫画标签获取问题
- 增强标签搜索功能
- 支持生成长图

**v0.3.7**
**v0.3.6**
- fix manhuagui标签搜索功能、封面图获取不到的问题
- 更换效率更高的pdf生成工具
- 添加manhuatai源

**v0.3.5**
- Added: 支持异步下载任务
- Added: 任务列表查询

**v0.3.4**
- Added: 支持聚合搜索接口
- Added: 支持标签过滤
- Improve: 接口返回值增加site、source_name字段
- Improve: 支持按站点配置代理
- Improve: 接入bumpveriosn版本管理
- Deprecate: 弃用图片下载缓存


onepiece/cli.py
import argparse
import os
import logging

from .comicbook import ComicBook
from .crawlerbase import CrawlerBase
from .utils import (
    parser_chapter_str,
    ensure_file_dir_exists
)
from .session import SessionMgr
from .worker import WorkerPoolMgr
from .utils.mail import Mail
from . import VERSION

logger = logging.getLogger(__name__)
HERE = os.path.abspath(os.path.dirname(__file__))

DEFAULT_DOWNLOAD_DIR = os.environ.get('ONEPIECE_DOWNLOAD_DIR') or 'download'
DEFAULT_MAIL_CONFIG_FILE = os.environ.get('ONEPIECE_MAIL_CONFIG_FILE') or ''
DEFAULT_DRIVER_TYPE = os.environ.get('ONEPIECE_DRIVER_TYPE') or 'Chrome'
DEFAULT_DRIVER_PATH = os.environ.get('ONEPIECE_DRIVER_PATH') or ''
DEFAULT_COOKIES_DIR = os.environ.get('ONEPIECE_COOKIES_DIR') or ''
DEFAULT_SESSION_DIR = os.environ.get('ONEPIECE_SESSION_DIR') or ''
DEFAULT_PROXY = os.environ.get('ONEPIECE_PROXY') or ''
DEFAULT_QUALITY = int(os.environ.get('ONEPIECE_QUALITY', 95))
DEFAULT_MAX_HEIGHT = int(os.environ.get('DEFAULT_MAX_HEIGHT', 20000))


def parse_args():
    """
    根据腾讯漫画id下载图片,默认下载海贼王最新一集。

    下载海贼王最新一集:
    python3 onepiece.py

    下载漫画 id=505430 最新一集:
    python3 onepiece.py -id 505430

    下载漫画 id=505430 所有章节:
    python3 onepiece.py -id 505430 -m all

    下载漫画 id=505430 第800集:
    python3 onepiece.py -id 505430 -c 800

    下载漫画 id=505430 倒数第二集:
    python3 onepiece.py -id 505430 -c -2

    下载漫画 id=505430 1到5集,7集，9到10集:
    python3 onepiece.py -id 505430 -i 1-5,7,9-10
    """

    parser = argparse.ArgumentParser(prog="onepiece")

    parser.add_argument('-id', '--comicid', type=str,
                        help="漫画id，如海贼王: 505430 (http://ac.qq.com/Comic/ComicInfo/id/505430)")
    parser.add_argument('--url', type=str,
                        help="漫画id，如海贼王: http://ac.qq.com/Comic/ComicInfo/id/505430")
    parser.add_argument('--url-file', type=str, help="漫画URL列表")

    parser.add_argument('--ext-name', type=str, help="如：番外篇、单行本等。具体得看站点支持哪些")

    parser.add_argument('--name', type=str, help="漫画名")

    parser.add_argument('-c', '--chapter', type=str, default="-1",
                        help="要下载的章节, 默认下载最新章节。如 -c 666 或者 -c 1-5,7,9-10")

    parser.add_argument('--worker', type=int, default=4, help="线程池数，默认开启4个线程池")

    parser.add_argument('--all', action='store_true',
                        help="是否下载该漫画的所有章节, 如 --all")

    parser.add_argument('--pdf', action='store_true',
                        help="是否生成pdf文件, 如 --pdf")
    parser.add_argument('--single-image', action='store_true',
                        help="是否拼接成一张图片, 如 --single-image")
    parser.add_argument('--quality', type=int, default=DEFAULT_QUALITY, help="生成长图的图片质量，最高质量100")
    parser.add_argument('--max-height', type=int, default=DEFAULT_MAX_HEIGHT, help="长图最大高度，最大高度65500")

    parser.add_argument('--login', action='store_true',
                        help="是否登录账号，如 --login")

    parser.add_argument('--mail', action='store_true',
                        help="是否发送pdf文件到邮箱, 如 --mail。需要预先配置邮件信息。\
                        可以参照config.ini.example文件，创建并修改config.ini文件")

    parser.add_argument('--receivers', type=str, help="邮件接收列表，多个以逗号隔开")
    parser.add_argument('--zip', action='store_true',
                        help="打包生成zip文件")

    parser.add_argument('--config', default=DEFAULT_MAIL_CONFIG_FILE, help="邮件配置文件路径")

    parser.add_argument('-o', '--output', type=str, default=DEFAULT_DOWNLOAD_DIR,
                        help="文件保存路径，默认保存在当前路径下的download文件夹")

    s = ' '.join(['%s(%s)' % (crawler.SITE, crawler.SOURCE_NAME) for crawler in ComicBook.CRAWLER_CLS_MAP.values()])
    site_help_msg = "数据源网站：支持 %s" % s

    parser.add_argument('-s', '--site', type=str, choices=ComicBook.CRAWLER_CLS_MAP.keys(),
                        help=site_help_msg)

    parser.add_argument('--verify', action='store_true',
                        help="verify")

    parser.add_argument('--driver-path', type=str, help="selenium driver", default=DEFAULT_DRIVER_PATH)

    parser.add_argument('--driver-type', type=str,
                        choices=CrawlerBase.SUPPORT_DRIVER_TYPE,
                        default=DEFAULT_DRIVER_TYPE,
                        help="支持的浏览器: {}. 默认为 {}".format(
                            ",".join(sorted(CrawlerBase.SUPPORT_DRIVER_TYPE)),
                            DEFAULT_DRIVER_TYPE)
                        )

    parser.add_argument('--session-path', type=str, help="读取或保存上次使用的session路径")
    parser.add_argument('--cookies-path', type=str, help="读取或保存上次使用的cookies路径")

    parser.add_argument('--latest-all', action='store_true', help="下载最近更新里的所有漫画")
    parser.add_argument('--latest-page', type=str, help="最近更新的页数，如1-10，默认第1页")

    parser.add_argument('--show-tags', action='store_true', help="展示当前支持的标签")
    parser.add_argument('--tag-all', action='store_true', help="下载标签里的所有漫画")
    parser.add_argument('--tag', type=str, help="标签id")
    parser.add_argument('--tag-page', type=str, help="标签页数，如1-10，默认第1页")

    parser.add_argument('--proxy', type=str,
                        help='设置代理，如 --proxy "socks5://user:pass@host:port"')

    parser.add_argument('-V', '--version', action='version', version=VERSION)
    parser.add_argument('--debug', action='store_true', help="debug")

    args = parser.parse_args()
    return args


def init_logger(level=None):
    level = level or logging.INFO
    logger = logging.getLogger()
    handler = logging.StreamHandler()
    formatter = logging.Formatter(
        "%(asctime)s %(name)s %(lineno)s [%(levelname)s] %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S'
    )
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    logger.setLevel(level)
    return logger


def download_main(comicbook, output_dir, ext_name=None, chapters=None,
                  is_download_all=None, is_gen_pdf=None, is_gen_zip=None,
                  is_single_image=None, quality=None, max_height=None, mail=None, receivers=None, is_send_mail=None):
    is_gen_pdf = is_gen_pdf or mail
    chapter_str = chapters or '-1'
    chapter_number_list = parser_chapter_str(chapter_str=chapter_str,
                                             last_chapter_number=comicbook.get_last_chapter_number(ext_name),
                                             is_all=is_download_all)
    for chapter_number in chapter_number_list:
        try:
            chapter = comicbook.Chapter(chapter_number, ext_name=ext_name)
            logger.info("正在下载 【{}】 {} 【{}】".format(
                comicbook.name, chapter.chapter_number, chapter.title))

            chapter_dir = chapter.save(output_dir=output_dir)
            logger.info("下载成功 %s", chapter_dir)
            if is_single_image:
                img_path = chapter.save_as_single_image(output_dir=output_dir, quality=quality, max_height=max_height)
                logger.info("生成长图 %s", img_path)
            if is_gen_pdf:
                pdf_path = chapter.save_as_pdf(output_dir=output_dir)
                logger.info("生成pdf文件 %s", pdf_path)

            if is_send_mail:
                mail.send(subject=os.path.basename(pdf_path),
                          content=None,
                          file_list=[pdf_path, ],
                          receivers=receivers)
            if is_gen_zip:
                zip_file_path = chapter.save_as_zip(output_dir=output_dir)
                logger.info("生成zip文件 %s", zip_file_path)
        except Exception:
            logger.exception('download comicbook error. site=%s comicid=%s chapter_number=%s',
                             comicbook.crawler.SITE, comicbook.crawler.comicid, chapter_number)


def download_latest_all(page_str, **kwargs):
    comicbook = kwargs.pop('comicbook')
    page_str = page_str or '1'
    for page in parser_chapter_str(page_str):
        for citem in comicbook.latest(page=page):
            next_comicbook = ComicBook(site=comicbook.crawler.SITE, comicid=citem.comicid)
            next_comicbook.start_crawler()
            echo_comicbook_desc(comicbook=next_comicbook, ext_name=kwargs.get('ext_name'))
            download_main(comicbook=next_comicbook, **kwargs)


def download_tag_all(tag, page_str, **kwargs):
    comicbook = kwargs.pop('comicbook')
    page_str = page_str or '1'
    for page in parser_chapter_str(page_str):
        for citem in comicbook.get_tag_result(tag=tag, page=page):
            next_comicbook = ComicBook(site=comicbook.crawler.SITE, comicid=citem.comicid)
            next_comicbook.start_crawler()
            echo_comicbook_desc(comicbook=next_comicbook, ext_name=kwargs.get('ext_name'))
            download_main(comicbook=next_comicbook, **kwargs)


def download_url_list(url_file, **kwargs):
    comicbook = kwargs.pop('comicbook')
    with open(url_file) as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith('#'):
                continue
            url = line
            site = ComicBook.get_site_by_url(url=url)
            comicid = ComicBook.get_comicid_by_url(site=site, url=url)
            if not site or not comicid:
                logger.info('Unknown url. url=%s', url)
                continue
            comicbook = ComicBook(site=site, comicid=comicid)
            comicbook.start_crawler()
            echo_comicbook_desc(comicbook=comicbook, ext_name=kwargs.get('ext_name'))
            download_main(comicbook=comicbook, **kwargs)


def show_tags(comicbook):
    msg_list = []
    for t1 in comicbook.get_tags():
        category = t1['category']
        t1_msg_list = []
        for t2 in t1['tags']:
            msg = '{name}={tag}'.format(name=t2['name'], tag=t2['tag'])
            t1_msg_list.append(msg)
        msg = '{}:\n{}'.format(category, '\t'.join(t1_msg_list))
        msg_list.append(msg)
    logger.info('支持的标签\n%s', '\n'.join(msg_list))


def echo_comicbook_desc(comicbook, ext_name=None):
    name = "{} {}".format(comicbook.name, ext_name) if ext_name else comicbook.name
    msg = ("{source_name} 【{name}】 更新至: {last_chapter_number:>03} "
           "【{last_chapter_title}】 数据来源: {source_url}").format(
        source_name=comicbook.source_name,
        name=name,
        last_chapter_number=comicbook.get_last_chapter_number(ext_name),
        last_chapter_title=comicbook.get_last_chapter_title(ext_name),
        source_url=comicbook.source_url)
    logger.info(msg)


def main():
    args = parse_args()

    if args.url:
        site = ComicBook.get_site_by_url(args.url)
        if not site:
            raise RuntimeError('Unknown url. url=%s' % args.url)
        comicid = ComicBook.get_comicid_by_url(site=site, url=args.url)
    else:
        site = args.site or 'qq'
        comicid = args.comicid

    session_path = args.session_path
    if not session_path and DEFAULT_SESSION_DIR:
        session_path = os.path.join(DEFAULT_SESSION_DIR, f'{site}.pickle')
    cookies_path = args.cookies_path
    if not cookies_path and DEFAULT_COOKIES_DIR:
        cookies_path = os.path.join(DEFAULT_COOKIES_DIR, f'{site}.json')

    loglevel = logging.DEBUG if args.debug else logging.INFO
    init_logger(level=loglevel)
    comicbook = ComicBook(site=site, comicid=comicid)
    # 设置代理
    proxy = args.proxy or os.environ.get('ONEPIECE_PROXY_{}'.format(site.upper())) or DEFAULT_PROXY
    if proxy:
        logger.info('set proxy. %s', proxy)
        SessionMgr.set_proxy(site=site, proxy=proxy)
    if args.verify:
        SessionMgr.set_verify(site=site, verify=True)

    WorkerPoolMgr.set_worker(worker=args.worker)
    CrawlerBase.DRIVER_PATH = args.driver_path
    CrawlerBase.DRIVER_TYPE = args.driver_type

    # 加载 session
    if session_path and os.path.exists(session_path):
        SessionMgr.load_session(site=site, path=session_path)
        logger.info('load session success. %s', session_path)

    # 加载cookies
    if cookies_path and os.path.exists(cookies_path):
        SessionMgr.load_cookies(site=site, path=cookies_path)
        logger.info('load cookies success. %s', cookies_path)

    if args.login:
        comicbook.crawler.login()

    if args.show_tags:
        show_tags(comicbook=comicbook)
        exit(0)

    if args.name:
        result = comicbook.search(name=args.name, limit=10)
        msg_list = []
        for item in result:
            msg_list.append("\ncomicid={}\t{}\tsource_url={}".format(
                item.comicid, item.name, item.source_url)
            )
        logger.info('\n%s', 'n'.join(msg_list))
        exit(0)

    if args.mail:
        is_send_mail = True
        mail = Mail.init(args.config)
    else:
        is_send_mail = False
        mail = None

    download_main_kwargs = dict(
        comicbook=comicbook,
        output_dir=args.output,
        chapters=args.chapter,
        is_download_all=args.all,
        is_gen_pdf=args.pdf,
        is_gen_zip=args.zip,
        is_single_image=args.single_image,
        quality=args.quality,
        max_height=args.max_height,
        mail=mail,
        ext_name=args.ext_name,
        is_send_mail=is_send_mail,
        receivers=args.receivers)

    if args.url_file:
        download_url_list(url_file=args.url_file, **download_main_kwargs)
    if args.latest_all:
        download_latest_all(page_str=args.latest_page, **download_main_kwargs)
    elif args.tag_all:
        download_tag_all(tag=args.tag, page_str=args.tag_page, **download_main_kwargs)
    else:
        logger.info("正在获取最新数据")
        comicbook.start_crawler()
        echo_comicbook_desc(comicbook=comicbook, ext_name=args.ext_name)
        download_main(**download_main_kwargs)

    # 保存 session
    if session_path:
        ensure_file_dir_exists(session_path)
        SessionMgr.export_session(site=site, path=session_path)
        logger.info("session saved. path={}".format(session_path))

    # 保存 cookies
    if cookies_path:
        ensure_file_dir_exists(cookies_path)
        SessionMgr.export_cookies(site=site, path=cookies_path)
        logger.info("cookies saved. path={}".format(cookies_path))


if __name__ == '__main__':
    main()
